plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,173 @@
+        when {
+            adjustedElements.isEmpty() -> {
+                showErrorHintByKey(
+                    project, editor, "cannot.refactor.no.expression",
+                    INTRODUCE_CONSTANT
+                )
+            }
+            else -> {
+                val options = ExtractionOptions(extractAsProperty = true)
+                val extractionData = ExtractionData(file, adjustedElements.toRange(), target, null, options)
+                ExtractionEngine(helper).run(editor, extractionData) {
+                    val property = it.declaration as KtProperty
+                    val descriptor = it.config.descriptor
+
+                    editor.caretModel.moveToOffset(property.textOffset)
+                    editor.selectionModel.removeSelection()
+                    if (editor.settings.isVariableInplaceRenameEnabled && !ApplicationManager.getApplication().isUnitTestMode) {
+                        with(PsiDocumentManager.getInstance(project)) {
+                            commitDocument(editor.document)
+                            doPostponedOperationsAndUnblockDocument(editor.document)
+                        }
+
+                        val introducer = KotlinInplacePropertyIntroducer(
+                            property = property,
+                            editor = editor,
+                            project = project,
+                            title = INTRODUCE_CONSTANT,
+                            doNotChangeVar = false,
+                            exprType = descriptor.returnType,
+                            extractionResult = it,
+                            availableTargets = listOf(ExtractionTarget.PROPERTY_WITH_GETTER)
+                        )
+                        introducer.performInplaceRefactoring(LinkedHashSet(getNameSuggestions(property) + descriptor.suggestedNames))
+                    } else {
+                        processDuplicatesSilently(it.duplicateReplacers, project)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun validateElements(elements: List<PsiElement>): String? {
+        val errorMessage = validateExpressionElements(elements)
+        return when {
+            errorMessage != null -> errorMessage
+            elements.any { it.isNotConst() } -> KotlinBundle.message("error.text.can.t.introduce.constant.for.this.expression.because.not.constant")
+            else -> null
+        }
+    }
+
+    private fun PsiElement.isNotConst() = !isConst()
+
+    private fun PsiElement.isConst(): Boolean {
+        if (this is KtBinaryExpression) {
+            var isConst = true
+            object : KotlinRecursiveElementVisitor() {
+                override fun visitKtElement(element: KtElement) {
+                    super.visitKtElement(element)
+                    if (element !is KtOperationReferenceExpression && element !is KtBinaryExpression) {
```

Comment:
I'll take a loot at it later.

Reply:
I added a field which contains all operators which are constant and used that to check like in `KtOperationReferenceExpression` Because the file grew a bit, I moved the `isConst` logic into a seperate util file